### PR TITLE
KAFKA-4899: Fix findbugs warnings in kafka-core

### PR DESCRIPTION
**Pull request by cmccabe**

### CODE REVIEW

**Comment by ijuma** [comment]
There's a test failing consistently.

File: core/src/main/scala/kafka/consumer/TopicCount.scala
Comment (ijuma) [inline core/src/main/scala/kafka/consumer/TopicCount.scala]: It seems a bit odd that this class implements `equals` but the `WildcardTopicCount` one doesn't.

File: gradle/findbugs-exclude.xml
Comment (ijuma) [inline gradle/findbugs-exclude.xml:56]: Hmm, do you know why this happens?

File: gradle/findbugs-exclude.xml
Comment (ijuma) [inline gradle/findbugs-exclude.xml:67]: Can we break this long line?

File: gradle/findbugs-exclude.xml
Comment (ijuma) [inline gradle/findbugs-exclude.xml:59]: Scala should use `valueOf` normally. Weird.

File: gradle/findbugs-exclude.xml
Comment (ijuma) [inline gradle/findbugs-exclude.xml:57]: This one is also a bit funny.

File: gradle/findbugs-exclude.xml
Comment (ijuma) [inline gradle/findbugs-exclude.xml:101]: I did notice that `case other: AnyRef` seems unnecessary. Is there more than that?

File: core/src/main/scala/kafka/metrics/KafkaMetricsReporter.scala
Comment (ijuma) [inline core/src/main/scala/kafka/metrics/KafkaMetricsReporter.scala]: We didn't remove this when cluster id was added because `KafkaMetricsReporter` is a pluggable interface and it wasn't clear to me if external users were using `KafkaMetricsReporter.ReporterStarted`. In other words, I assumed this class to be public since `KafkaMetricsReporter` is public.

**Review by ijuma — COMMENTED** [review]
Thanks for the PR. Did a quick pass and left some comments.

File: gradle/findbugs-exclude.xml
Comment (cmccabe) [reply gradle/findbugs-exclude.xml:56]: Findbugs scans generated bytecode, not actual source code files.  My guess is that the bytecode Scala generates sometimes checks for nulls even when they are not possible, or invokes new String(String).  All the times that I checked up on these issues in scala code, they were in the generated code, not in the code that the programmer wrote.

File: core/src/main/scala/kafka/metrics/KafkaMetricsReporter.scala
Comment (cmccabe) [reply core/src/main/scala/kafka/metrics/KafkaMetricsReporter.scala]: This boolean never gets set to true, though.  Is there really any chance this is useful?

File: core/src/main/scala/kafka/consumer/TopicCount.scala
Comment (cmccabe) [reply core/src/main/scala/kafka/consumer/TopicCount.scala]: Good point.  I don't think anywhere is actually using this equals method... let me try re-running the tests with an equals() method that throws an exception.

File: core/src/main/scala/kafka/metrics/KafkaMetricsReporter.scala
Comment (ijuma) [reply core/src/main/scala/kafka/metrics/KafkaMetricsReporter.scala]: Why does it never get set to `true`? If you configure `KafkaMetricReporters`, it should be set to `true` right?

Comment (cmccabe) [reply core/src/main/scala/kafka/metrics/KafkaMetricsReporter.scala]: Sorry, I was looking at a different version of the code.  The existing version of the code in trunk does set it.  I suppose we should keep this as-is.

File: core/src/main/scala/kafka/consumer/TopicCount.scala
Comment (jozanek) [inline core/src/main/scala/kafka/consumer/TopicCount.scala:109]: I do not get this one. Why would you ever wanted to override equals for any scala "object"?

File: core/src/main/scala/kafka/javaapi/FetchRequest.scala
Comment (jozanek) [inline core/src/main/scala/kafka/javaapi/FetchRequest.scala:68]: I think you can achieve the same by making FetchRequest case class, then you can delete equals and also hashCode.

File: core/src/main/scala/kafka/log/Log.scala
Comment (jozanek) [inline core/src/main/scala/kafka/log/Log.scala]: None is returned to no val, I think this is unnecessary. What do you want to do if directory already exist? Do you want to log in somewhere? Or maybe just ignore it?

File: core/src/main/scala/kafka/tools/ImportZkOffsets.scala
Comment (jozanek) [inline core/src/main/scala/kafka/tools/ImportZkOffsets.scala]: Maybe offtopic, but why do we not use scala.io.Source here instead of this *Reader stuff?

File: core/src/main/scala/kafka/tools/ImportZkOffsets.scala
Comment (jozanek) [inline core/src/main/scala/kafka/tools/ImportZkOffsets.scala]: If possible I vote for rewriting this method to use only vals.

**Comment by jozanek** [comment]
Hey guys, I accidentally found this PR and added my 2 cents. I have not seen output from findbugs, so maybe they are inaccurate.

File: gradle/findbugs-exclude.xml
Comment (cmccabe) [reply gradle/findbugs-exclude.xml:67]: Good idea.  I found a way of doing this by using the \<or\> construct here.

File: core/src/main/scala/kafka/javaapi/FetchRequest.scala
Comment (cmccabe) [reply core/src/main/scala/kafka/javaapi/FetchRequest.scala:68]: I think that's a bigger refactoring that is worth considering in its own JIRA.  This is just a minimal set of changes to get findbugs to be clean.

File: core/src/main/scala/kafka/log/Log.scala
Comment (cmccabe) [reply core/src/main/scala/kafka/log/Log.scala]: We want to ignore cases where the directory already exists.  I will change it to => {}

File: core/src/main/scala/kafka/tools/ImportZkOffsets.scala
Comment (cmccabe) [reply core/src/main/scala/kafka/tools/ImportZkOffsets.scala]: Would you like to follow a follow-on JIRA to change this to scala.io.Source?  It seems unrelated to fixing the findbugs issue, but perhaps it is a good idea.

File: core/src/main/scala/kafka/tools/ImportZkOffsets.scala
Comment (cmccabe) [reply core/src/main/scala/kafka/tools/ImportZkOffsets.scala]: See above (sounds like a good candidate for a cleanup JIRA)

File: core/src/main/scala/kafka/log/Log.scala
Comment (jozanek) [reply core/src/main/scala/kafka/log/Log.scala]: Then you can do just Try { Files.createDirectory(dir.toPath) }

File: core/src/main/scala/kafka/tools/ImportZkOffsets.scala
Comment (jozanek) [reply core/src/main/scala/kafka/tools/ImportZkOffsets.scala]: Sure, I can do that.

File: gradle/findbugs-exclude.xml
Comment (cmccabe) [reply gradle/findbugs-exclude.xml:101]: The warning is:
                                                                                                                                                                                                                                                                                          

> BC instanceof will always return true for all non-null values in kafka.utils.Json$$anonfun$encode$1.apply(Tuple2), since all scala.Tuple2<java.lang.Object,java.lang.Object> are instances of scala.Tuple2

Seems like a pretty clear-cut instance of generated code triggering a spurious warning.

File: core/src/main/scala/kafka/consumer/TopicCount.scala
Comment (cmccabe) [reply core/src/main/scala/kafka/consumer/TopicCount.scala:109]: We are trying to ascertain whether we can get rid of `WildcardTopicCount#equals`, by adding a an equals method in the superclass that always throws an exception.  If the unit tests pass with this configuration, we should be able to get rid of the other one.

File: core/src/main/scala/kafka/javaapi/FetchRequest.scala
Comment (ijuma) [reply core/src/main/scala/kafka/javaapi/FetchRequest.scala:68]: Let's keep changes to the Scala request classes as small as possible. The goal is to delete them as soon as we remove the old clients.

**Comment by cmccabe** [comment]
Thanks for the reviews, @ijuma  and @jozi-k.  I made some revisions, so check it out.

@ijuma, the exclusions for scala code are very necessary.  Findbugs does apparently trip up on some generated Scala code, and generate thousands of spurious warnings.

@jozi-k, there are tons of style cleanups we could do here.  I think we should focus on just getting findbugs clean here, and fixing things that are clearly bugs.  Then we can file follow-on JIRAs for any style cleanups that we feel like doing later.  This is part of the bigger effort of getting AK findbugs-clean so that we can enable findbugs-warnings-as-errors to prevent regressions.

@ijuma, you are right that the fact that TopicCount has no equals method, but one of its subclasses does, is a little suspicious.  I am re-running the unit tests with an equals method that throws an exception to see if the exception is ever thrown.

File: core/src/main/scala/kafka/tools/ImportZkOffsets.scala
Comment (ijuma) [reply core/src/main/scala/kafka/tools/ImportZkOffsets.scala]: `Source` is a class designed to read source files and it had a few pitfalls so the recommendation is generally to stay away (see http://stackoverflow.com/questions/4458864/whats-the-right-way-to-use-scala-io-source for one example).

File: core/src/main/scala/kafka/consumer/TopicCount.scala
Comment (jozanek) [reply core/src/main/scala/kafka/consumer/TopicCount.scala:109]: `WildcardTopicCount` does not override `equals` method at all. Generally, it is quite bad idea to throw exception during object comparison. Rather, I would get rid of `StaticTopicCount#equals` by making it case class.

File: gradle/findbugs-exclude.xml
Comment (ijuma) [reply gradle/findbugs-exclude.xml:56]: Yes, I'm aware of how findBugs works and that it flags issues in the code generated by the Scala compiler. Sorry for not making that clear. There were a few new ones that I didn't come across last time I did this exercise and I was curious about the code pattern that triggered in case you had already tracked it down. No worries.

File: core/src/main/scala/kafka/tools/ImportZkOffsets.scala
Comment (jozanek) [reply core/src/main/scala/kafka/tools/ImportZkOffsets.scala]: In our case we just read and parse relatively small input file, so it is fine to use it. Much better than doing unnecessary mutation, or checking for null.

**Comment by jozanek** [comment]
I agree that only clear bugs should be resolved first. As per TopicCount, you extend the trait, not the object so equals in object is useless in this case. Consider making it case class instead.

File: core/src/main/scala/kafka/javaapi/FetchRequest.scala
Comment (jozanek) [reply core/src/main/scala/kafka/javaapi/FetchRequest.scala:68]: Ah, I see. Does it mean all scala code will be eventually replaced in the future?

File: core/src/main/scala/kafka/consumer/TopicCount.scala
Comment (ijuma) [reply core/src/main/scala/kafka/consumer/TopicCount.scala:109]: @jozi-k, we don't intend to merge this change. We are just checking if `equals` is ever invoked during the tests.

File: core/src/main/scala/kafka/javaapi/FetchRequest.scala
Comment (ijuma) [reply core/src/main/scala/kafka/javaapi/FetchRequest.scala:68]: The current plan is that code that is used only by the broker and is not public API will remain in Scala. Code that is common or is public API will be in Java.

File: core/src/main/scala/kafka/log/Log.scala
Comment (ijuma) [inline core/src/main/scala/kafka/log/Log.scala]: Hmm, shouldn't this be `Files.createDirectories(dir.toPath)`? That seems to be the equivalent of `mkdirs` and it doesn't throw an exception if the directory already exists. There are a few others like this.

By the way, this fixes KAFKA-3940.

File: core/src/main/scala/kafka/log/Log.scala
Comment (ijuma) [inline core/src/main/scala/kafka/log/Log.scala:227]: One thing that is not clear to me is if we want to just fail if we fail to delete or if we want to try to delete all the files and throw the exception at the end. What do you think?

File: core/src/main/scala/kafka/log/LogCleaner.scala
Comment (ijuma) [inline core/src/main/scala/kafka/log/LogCleaner.scala]: What's this (and other similar cases) for?

File: core/src/main/scala/kafka/log/LogManager.scala
Comment (ijuma) [inline core/src/main/scala/kafka/log/LogManager.scala]: No `;` required.

File: core/src/main/scala/kafka/log/LogSegment.scala
Comment (ijuma) [inline core/src/main/scala/kafka/log/LogSegment.scala]: Maybe we should reuse the same `FileTime`? Also a nit: `toPath()` should just be `toPath`, there are a few cases like that in the PR.

File: core/src/main/scala/kafka/server/ClientQuotaManager.scala
Comment (ijuma) [inline core/src/main/scala/kafka/server/ClientQuotaManager.scala:132]: What is the reason for this?

File: core/src/main/scala/kafka/server/OffsetCheckpoint.scala
Comment (ijuma) [inline core/src/main/scala/kafka/server/OffsetCheckpoint.scala]: `case _: FileAlreadyExistsException => ()`. There are a few others like this one.

File: core/src/main/scala/kafka/tools/MirrorMaker.scala
Comment (ijuma) [inline core/src/main/scala/kafka/tools/MirrorMaker.scala]: Nit: no space before `:`. Same below.

File: core/src/main/scala/kafka/tools/MirrorMaker.scala
Comment (ijuma) [inline core/src/main/scala/kafka/tools/MirrorMaker.scala]: This is to protect against spurious wake ups, right? `iter` is not null after initialisation, so it doesn't seem like it works for that purpose. Seems like `immediateCommitRequested` could be that if we do something like:

```scala
if (iter != null) {
  immediateCommitRequested = true
  while (!immediateCommitRequested)
    this.wait()
```

But please check if this makes sense.

**Review by ijuma — COMMENTED** [review]
Thanks for the updates. I left a few more comments. Sorry for the mix between `single` and `batch` comments. Halfway, I realise that there were a few more things I had not noticed.

**Comment by ijuma** [comment]
Also, one conflict needs to be resolved.

File: core/src/main/scala/kafka/consumer/TopicCount.scala
Comment (jozanek) [reply core/src/main/scala/kafka/consumer/TopicCount.scala:109]: Understood.

File: core/src/main/scala/kafka/log/Log.scala
Comment (cmccabe) [reply core/src/main/scala/kafka/log/Log.scala:227]: My gut feeling would be to keep it simple for now, and have more elaborate error handling later if we find scenarios where it would have helped.  I think it's rare to find a filesystem that has undeletable files that is still usable in practice (but maybe there is a scenario I haven't considered?).

File: core/src/main/scala/kafka/log/LogCleaner.scala
Comment (cmccabe) [reply core/src/main/scala/kafka/log/LogCleaner.scala]: The two-argument form of `CountDownLatch#await` returns `true` if the latch was decremented, and `false` if it timed out.  This code doesn't check this return value.  It's not a bug, because there is code later on which is calling `checkDone` to determine if the cleaner should exit.  So we don't need the return value from `await`.  But it triggers an "ignored return value" findbugs warning.  Using `val _ = ...await...` suppresses this.

File: gradle/findbugs-exclude.xml
Comment (cmccabe) [reply gradle/findbugs-exclude.xml:56]: Sorry, I wasn't trying to be pedantic :)  It seemed like a good idea to mention it for the other folks casually reading this.  I do wish we could enable some of these warnings, but there was just way too much noise.  I can give more examples if that would be helpful.

File: core/src/main/scala/kafka/server/ClientQuotaManager.scala
Comment (cmccabe) [reply core/src/main/scala/kafka/server/ClientQuotaManager.scala:132]: findbugs warns about starting threads in a class's constructor, *unless* that class is final.  The reason for this is that if it is ever subclassed, the thread would be started in the superclass's constructor, before any of the derived fields had been initialized, which could lead to surprising results.  It seemed easier to just make the class final than to add a suppression or refactor the class to not do that.

File: core/src/main/scala/kafka/tools/MirrorMaker.scala
Comment (cmccabe) [reply core/src/main/scala/kafka/tools/MirrorMaker.scala]: Thanks, that's a good catch.  I added a counter to make sure that we don't get stuck waiting for someone else's invocation of `requestAndWaitForCommit` (which would reset the `immediateCommitRequest` variable back to true)

**Comment by cmccabe** [comment]
retest this please

File: core/src/main/scala/kafka/consumer/TopicCount.scala
Comment (ijuma) [reply core/src/main/scala/kafka/consumer/TopicCount.scala:109]: Can we remove this code now?

File: core/src/main/scala/kafka/log/Log.scala
Comment (ijuma) [inline core/src/main/scala/kafka/log/Log.scala:287]: Why is this one not `deleteIfExists`?

File: core/src/main/scala/kafka/log/LogCleaner.scala
Comment (ijuma) [reply core/src/main/scala/kafka/log/LogCleaner.scala]: I see. Unfortunately, this is a bit of an implementation detail. The scala compiler could simply remove the local variable from the bytecode since it's not used (I think it does that with -optimise) and that would trigger the warning again. Maybe we want to disable the warning instead?

Comment (cmccabe) [reply core/src/main/scala/kafka/log/LogCleaner.scala]: Good point.  I will add a suppression.

File: core/src/main/scala/kafka/log/Log.scala
Comment (cmccabe) [reply core/src/main/scala/kafka/log/Log.scala:287]: This is inside an `if` block where we checked that `indexFile` exists.

**Comment by cmccabe** [comment]
Looks like Jenkins went belly up?
```
GitHub pull request #2687 of commit 1a7fdd87c1b97bf210f561eb00c79185af82f3af, no merge conflicts.
Setting status of 1a7fdd87c1b97bf210f561eb00c79185af82f3af to PENDING with url https://builds.apache.org/job/kafka-pr-jdk8-scala2.11/2656/ and message: 'Build started sha1 is merged.'
Using context: JDK 8 and Scala 2.11
[EnvInject] - Loading node environment variables.
ERROR: SEVERE ERROR occurs
org.jenkinsci.lib.envinject.EnvInjectException: hudson.remoting.ChannelClosedException: channel is already closed
	at org.jenkinsci.plugins.envinject.service.EnvironmentVariablesNodeLoader.gatherEnvironmentVariablesNode(EnvironmentVariablesNodeLoader.java:79)
	at org.jenkinsci.plugins.envinject.EnvInjectListener.loadEnvironmentVariablesNode(EnvInjectListener.java:80)
	at org.jenkinsci.plugins.envinject.EnvInjectListener.setUpEnvironment(EnvInjectListener.java:42)
	at hudson.model.AbstractBuild$AbstractBuildExecution.createLauncher(AbstractBuild.java:572)
	at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:492)
	at hudson.model.Run.execute(Run.java:1728)
	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
	at hudson.model.ResourceController.execute(ResourceController.java:98)
	at hudson.model.Executor.run(Executor.java:404)
Caused by: hudson.remoting.ChannelClosedException: channel is already closed
	at hudson.remoting.Channel.send(Channel.java:604)
	at hudson.remoting.Request.call(Request.java:130)
	at hudson.remoting.Channel.call(Channel.java:821)
	at hudson.FilePath.act(FilePath.java:1080)
	at org.jenkinsci.plugins.envinject.service.EnvironmentVariablesNodeLoader.gatherEnvironmentVariablesNode(EnvironmentVariablesNodeLoader.java:48)
	... 8 more
Caused by: java.io.IOException
	at hudson.remoting.Channel.close(Channel.java:1284)
	at hudson.slaves.ChannelPinger$1.onDead(ChannelPinger.java:115)
	at hudson.remoting.PingThread.ping(PingThread.java:130)
	at hudson.remoting.PingThread.run(PingThread.java:86)
Caused by: java.util.concurrent.TimeoutException: Ping started at 1491240968010 hasn't completed by 1491241208011
	... 2 more
ERROR: Step ‘Publish JUnit test result report’ failed: no workspace for kafka-pr-jdk8-scala2.11 #2656
[EnvInject] - [ERROR] - SEVERE ERROR occurs: channel is already closed
Adding one-line test results to commit status...
Setting status of 1a7fdd87c1b97bf210f561eb00c79185af82f3af to FAILURE with url https://builds.apache.org/job/kafka-pr-jdk8-scala2.11/2656/ and message: 'Build finished. No test results found.'
Using context: JDK 8 and Scala 2.11
```

**Comment by ijuma** [comment]
retest this please

**Comment by ijuma** [comment]
Can you please rebase this one?

File: core/src/main/scala/kafka/tools/MirrorMaker.scala
Comment (ijuma) [inline core/src/main/scala/kafka/tools/MirrorMaker.scala:533]: Would the following not work?

```scala
          immediateCommitRequested = true
          do {
            this.wait()
          } while (immediateCommitRequested)
```

File: core/src/main/scala/kafka/consumer/TopicCount.scala
Comment (ijuma) [inline core/src/main/scala/kafka/consumer/TopicCount.scala:107]: I removed this before merging.

File: core/src/main/scala/kafka/server/checkpoints/CheckpointFile.scala
Comment (ijuma) [inline core/src/main/scala/kafka/server/checkpoints/CheckpointFile.scala:39]: I fixed the formatting of this and the other similar case before merging.

**Review by ijuma — APPROVED** [review]
LGTM, merged to trunk after minor changes (see my comments below).

Also, I left one question regarding the MirrorMaker change. If the simpler approach I suggested works, we should do that in a follow-up.

**Pull request closed without merging**
